### PR TITLE
Watch for changes in the registration secret

### PR DESCRIPTION
**Pull request by mgianatagh**
# Description

Watch for changes in the registration secret.  

- If secret is deleted stop syncing multi-cluster objects
- If the secret changes, re-establish the connection to the admin cluster with the new values

Fixes VZ-2228

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/Makefile [hunk 1]
```diff
@@ -166,14 +166,14 @@ export VERRAZZANO_APP_OP_IMAGE=${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}
 .PHONY: setup-cluster
 setup-cluster: create-cluster
 	echo 'Load Docker image for the OAM runtime...'
-	docker pull ${OAM_RUNTIME_IMAGE}
-	kind load docker-image --name ${CLUSTER_NAME} ${OAM_RUNTIME_IMAGE}
+	time docker pull ${OAM_RUNTIME_IMAGE}
```
Comment (pfmackin) [inline application-operator/Makefile:169]: i assume this is not directly related to the PR

File: application-operator/mcagent/mcagent.go [hunk 2]
```diff
@@ -25,82 +27,99 @@ func StartAgent(client client.Client, log logr.Logger) {
...
-	log.Info(fmt.Sprintf("Found secret named %s in namespace %s for cluster named %q", secret.Name, secret.Namespace, managedClusterName))
+// ProcessAgentThread - process one iteration of the agent thread
+func (s *Syncer) ProcessAgentThread() error {
+	secret := corev1.Secret{}
 
-	// Create the client for accessing the admin cluster
-	adminClient, err := getAdminClient(&secret)
+	// Get the secret
+	err := s.LocalClient.Get(context.TODO(), types.NamespacedName{Name: constants.MCAgentSecret, Namespace: constants.VerrazzanoSystemNamespace}, &secret)
+	if err != nil {
```
Comment (pfmackin) [inline application-operator/mcagent/mcagent.go:57]: should check for errrors.isNotFound

File: application-operator/Makefile (hunk 1 above)
Comment (mgianatagh) [reply application-operator/Makefile:169]: Yes, this is in response to a conversation I had with Tony.  I had an anomaly where the integ tests took 45 minutes to run.  He suggested adding timings so we can better diagnose better the next time.

File: application-operator/mcagent/mcagent.go (hunk 2 above)
Comment (mgianatagh) [reply application-operator/mcagent/mcagent.go:57]: that is happening on line 58 via the call to clusters.IgnoreNotFoundWithLog

File: application-operator/mcagent/mcagent.go [hunk 3]
```diff
@@ -25,82 +27,99 @@ func StartAgent(client client.Client, log logr.Logger) {
...
+		if clusters.IgnoreNotFoundWithLog("secret", err, s.Log) == nil && s.AgentSecretFound {
+			s.Log.Info(fmt.Sprintf("the secret %s in namespace %s was deleted", constants.MCAgentSecret, constants.VerrazzanoSystemNamespace))
```
Comment (desagar) [inline application-operator/mcagent/mcagent.go:59]: would this log about secret deletion also print for every iteration until the secret is first created?

File: application-operator/mcagent/mcagent.go [hunk 4]
```diff
@@ -25,82 +27,99 @@ func StartAgent(client client.Client, log logr.Logger) {
...
-		ManagedClusterName: managedClusterName,
-		Context:            context.TODO(),
-		ProjectNamespaces:  []string{},
+	// Remember the secret had been found in order to notice if it gets deleted
+	s.AgentSecretFound = true
+
+	// The cluster secret exists - log the cluster name only if it changes
+	managedClusterName := string(secret.Data[constants.ClusterNameData])
+	if managedClusterName != s.ManagedClusterName {
+		s.Log.Info(fmt.Sprintf("Found secret named %s in namespace %s for cluster named %q", secret.Name, secret.Namespace, managedClusterName))
```
Comment (desagar) [inline application-operator/mcagent/mcagent.go:75]: I'd prefer if this log said that the managed cluster name changed from X to Y

File: application-operator/mcagent/mcagent_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			secret.ObjectMeta = validSecret.ObjectMeta
+			secret.Data = validSecret.Data
+			return nil
+		})
+
+	// Admin Cluster - expect call to list VerrazzanoProject objects - return an empty list
+	adminMock.EXPECT().
+		List(gomock.Any(), &clustersv1alpha1.VerrazzanoProjectList{}, gomock.Not(gomock.Nil())).
+		DoAndReturn(func(ctx context.Context, list *clustersv1alpha1.VerrazzanoProjectList, opts ...*client.ListOptions) error {
+			return fmt.Errorf("no project resources found %s", exec.ErrNotFound)
```
Comment (desagar) [inline application-operator/mcagent/mcagent_test.go:61]: Does this work as expected? 
I don't think exec.ErrNotFound is the right kind of NotFound error. There is a Kubernetes api function called NewNotFound which is used in some of the MC controller tests. This is what is interpreted as a not found error in the clusters pkg function to ignore not found errors.

File: application-operator/mcagent/mcagent_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+	// Admin cluster mocks
+	adminMocker := gomock.NewController(t)
+	adminMock := mocks.NewMockClient(adminMocker)
+
+	// Managed Cluster - expect call to get the cluster registration secret.
+	mcMock.EXPECT().
+		Get(gomock.Any(), types.NamespacedName{Namespace: constants.VerrazzanoSystemNamespace, Name: constants.MCAgentSecret}, gomock.Not(gomock.Nil())).
+		DoAndReturn(func(ctx context.Context, name types.NamespacedName, secret *corev1.Secret) error {
+			return fmt.Errorf("no project resources found %s", exec.ErrNotFound)
```
Comment (desagar) [inline application-operator/mcagent/mcagent_test.go:101]: same comment as above, plus a copy/paste error - this err msg should say something like agent secret not found

File: application-operator/mcagent/mcagent.go (hunk 3 above)
Comment (mgianatagh) [reply application-operator/mcagent/mcagent.go:59]: No, because "s.AgentSecretFound" would initially be false and it is part of the "and" condition

Comment (desagar) [reply application-operator/mcagent/mcagent.go:59]: thanks didn't notice that.

File: application-operator/mcagent/mcagent.go (hunk 4 above)
Comment (mgianatagh) [reply application-operator/mcagent/mcagent.go]: changed in latest push

File: application-operator/mcagent/mcagent_test.go (hunk 5 above)
Comment (mgianatagh) [reply application-operator/mcagent/mcagent_test.go]: It seemed to work, but was probably the wrong choice.  I changed to use NewNotFound

File: application-operator/mcagent/mcagent_test.go (hunk 6 above)
Comment (mgianatagh) [reply application-operator/mcagent/mcagent_test.go]: I changed to use NewNotFound

**Review by desagar — APPROVED** [review]
